engine/include/cubos/engine/collisions/raycast.hpp

```diff
@@ -0,0 +1,255 @@
+
+#include <cubos/core/ecs/system/fetcher.hpp>
+#include <cubos/core/geom/box.hpp>
+
+#include <cubos/engine/collisions/collider.hpp>
+#include <cubos/engine/collisions/shapes/box.hpp>
+#include <cubos/engine/collisions/shapes/capsule.hpp>
+#include <cubos/engine/prelude.hpp>
+#include <cubos/engine/transform/local_to_world.hpp>
+#include <cubos/engine/transform/position.hpp>
+
+namespace cubos::engine
+{
+    struct Hit
+    {
+        Entity entity;
+        glm::vec3 point;
+    };
+
+    struct Ray
+    {
+        glm::vec3 origin, direction;
+    };
+
+    class Raycast
+    {
+    public:
+        Raycast(Query<Entity, const LocalToWorld&, const BoxCollisionShape&> boxes,
+                Query<Entity, const LocalToWorld&, const CapsuleCollisionShape&, const Position&> capsules)
+            : mBoxes{std::move(boxes)}
+            , mCapsules{std::move(capsules)} {};
+
+        Opt<Hit> fire(Ray ray)
+        {
+            Opt<Hit> hitMax;
+            float minScalar = std::numeric_limits<float>::infinity();
+
+            // normalize the ray
+            ray.direction = glm::normalize(ray.direction);
+
+            for (auto [entity, localToWorld, shape] : mBoxes)
+            {
+                auto worldToLocal = localToWorld.inverse();
+                Ray localRay = {glm::vec3(worldToLocal * glm::vec4(ray.origin, 1.0F)),
+                                glm::normalize(glm::vec3(worldToLocal * glm::vec4(ray.direction, 0.0F)))};
+                float scalar = intersects(localRay, shape.box);
+
+                if (scalar <= minScalar && scalar >= 0.0F)
+                {
+                    Hit hit;
+                    hit.entity = entity;
+                    hit.point = glm::vec3(ray.origin + scalar * ray.direction);
+                    hitMax.replace(hit);
+                    minScalar = scalar;
+                }
+            }
+
+            for (auto [entity, localToWorld, shape, position] : mCapsules)
+            {
+                auto worldToLocal = localToWorld.inverse();
+                Ray localRay = {glm::vec3(worldToLocal * glm::vec4(ray.origin, 1.0F)),
+                                glm::normalize(glm::vec3(worldToLocal * glm::vec4(ray.direction, 0.0F)))};
+                float scalar = intersects(localRay, worldToLocal, shape.capsule, position);
+                if (scalar <= minScalar && scalar >= 0.0F)
+                {
+                    Hit hit;
+                    hit.entity = entity;
+                    hit.point = glm::vec3(ray.origin + scalar * ray.direction);
+                    hitMax.replace(hit);
+                    minScalar = scalar;
+                }
+            }
+
+            return hitMax;
+        }
+
+    private:
+        Query<Entity, const LocalToWorld&, const BoxCollisionShape&> mBoxes;
+        Query<Entity, const LocalToWorld&, const CapsuleCollisionShape&, const Position&> mCapsules;
+
+        static float intersects(Ray ray, cubos::core::geom::Box box)
+        {
```

Comment:
Should be moved to the .cpp file entirely. It doesn't need to be a method of the class, you can just keep in the source file as is. I.e.


```cpp
#include <...>

static float intersects(...)
{
...
}
```